src/Servers/Kestrel/Core/src/Internal/Http3/Http3OutputProducer.cs

```diff
@@ -0,0 +1,413 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers;
+using Microsoft.Extensions.Logging;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Internal;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    internal class Http3OutputProducer : IHttpOutputProducer, IHttpOutputAborter
+    {
+        private readonly int _streamId;
+        private readonly Http3FrameWriter _frameWriter;
+        private readonly TimingPipeFlusher _flusher;
+        private readonly IKestrelTrace _log;
+        private readonly MemoryPool<byte> _memoryPool;
+        private readonly Http3Stream _stream;
+        private readonly PipeWriter _pipeWriter;
+        private readonly PipeReader _pipeReader;
+        private readonly object _dataWriterLock = new object();
+        private readonly ValueTask<FlushResult> _dataWriteProcessingTask;
+        private bool _startedWritingDataFrames;
+        private bool _completed;
+        private bool _disposed;
+        private bool _suffixSent;
+        private IMemoryOwner<byte> _fakeMemoryOwner;
+
+        public Http3OutputProducer(
+             int streamId,
+             Http3FrameWriter frameWriter,
+             MemoryPool<byte> pool,
+             Http3Stream stream,
+             IKestrelTrace log)
+        {
+            _streamId = streamId;
+            _frameWriter = frameWriter;
+            _memoryPool = pool;
+            _stream = stream;
+            _log = log;
+
+            var pipe = CreateDataPipe(pool);
+
+            _pipeWriter = pipe.Writer;
+            _pipeReader = pipe.Reader;
+
+            // No need to pass in timeoutControl here, since no minDataRates are passed to the TimingPipeFlusher.
+            // The minimum output data rate is enforced at the connection level by Http2FrameWriter.
```

Comment:
This comment probably needs to be updated. This seems very similar to Http2OutputProducer. It would be nice if we could share some more of the logic.

Reply:
Probably, however, I'd be worried about timeout control and minDataRates (whole connection vs stream limits). That will probably require further discussion as well.